### ISSUE TITLE
Add scope attributes

### ISSUE THREAD

**Issue by jack-berg** [open-telemetry/opentelemetry-java#4695]
Scopes have [attributes](https://github.com/open-telemetry/opentelemetry-specification/pull/2579). We need to extend our data model to support this, which is particularly important [metrics](https://github.com/open-telemetry/opentelemetry-specification/pull/2702) and for the [event api](https://github.com/open-telemetry/opentelemetry-specification/pull/2676) where a scope attribute called `event.domain` seems likely to be required.

There are a number of small things to do for scope attributes to be supported:
- [ ] Extend `InstrumentationScopeInfo` with attributes
- [ ] Extend `TracerBuilder` to optionally specify attributes
- [ ] Extend `MeterBuilder` to optionally specify attributes
- [ ] Extend `LogEmitterBuilder` to optionally specify attributes
- [ ] Serialize scope attributes in OTLP
- [ ] Serialize scope attributes in zipkin
- [ ] Serialize scope attributes in prometheus

**Comment by jack-berg** [comment]
Serialization of jaeger and zipkin scope attributes is blocked on spec issue [#2744](https://github.com/open-telemetry/opentelemetry-specification/issues/2744).

**Comment by chukunx** [comment]
Hey @jack-berg and folks, now that spec issue 2744 has been resolved, do we have plans to implement this request? Anything else to consider beyond what’s defined in the spec?

**Comment by patpatpat123** [comment]
# Scope attributes on `TracerBuilder` / `MeterBuilder` / `LoggerBuilder`

A worked-out implementation for [open-telemetry/opentelemetry-java#4695 — "Add scope attributes"](https://github.com/open-telemetry/opentelemetry-java/issues/4695).

Everything below has been written and verified against `main` at `d948e130e` (1.66.0-SNAPSHOT). Nothing has been committed or submitted anywhere — this is a proposal for you to review, argue with, and paste into your IDE if you like it.

---

## TL;DR

Add a `default setAttributes(Attributes)` to the three stable builder interfaces, wire the three SDK builders to pass it through instead of hardcoding `Attributes.empty()`, and — while we're here — turn `ComponentRegistry`'s "behavior is undefined" into behaviour that's defined *and* tells you when it bites you.

14 files: 7 production, 5 test, 2 meta (CHANGELOG + apidiff). Full `./gradlew build` is green.

---

## Why this is worth picking up now

The issue has been open since August 2022 and the checklist hasn't moved in four years, so it's fair to ask what's actually different. Three things:

**1. The spec compliance matrix marks this as required, and Java is the only language failing it.**

From [`spec-compliance-matrix.md`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md):

```
| Feature                            | Optional | Go | Java | JS | Python | Ruby | Erlang | PHP | Rust | C++ | .NET | Swift | Kotlin |
| Get a Tracer with scope attributes |          | +  |  -   |    |   +    |      |        |  +  |  +   |  +  |  +   |       |   +    |
```

A blank `Optional` column means required. Go, Python, PHP, Rust, C++, .NET and Kotlin have all shipped it. Java carries an explicit `-`.

**2. The stated blocker is gone, and was never the real one anyway.**

[spec#2744](https://github.com/open-telemetry/opentelemetry-specification/issues/2744) — the thing Jack Berg flagged in 2022 — closed as COMPLETED in April 2026. It's also moot: it only gated the Zipkin and Jaeger serialization items, and neither exporter exists in this repo anymore (Zipkin was deleted in 1.65.0 via [#8677](https://github.com/open-telemetry/opentelemetry-java/pull/8677)). Those two checkboxes on the issue are unresolvable as written and should probably just be struck so the issue can actually close.

**3. Someone asked.**

chukunx asked on 2026-06-30 whether there were plans to implement it now that 2744 is resolved. That question is still unanswered. The 2022 decision to close [PR #4711](https://github.com/open-telemetry/opentelemetry-java/pull/4711) was explicitly *"I think we should close this until users ask for it"* — so that gate has arguably been met.

There's also a smaller thing that bugs me: `TracerBuilder`'s class Javadoc already ships this sentence today —

> *"These identifying fields, along with attributes, combine to form the instrumentation scope"*

— describing something the interface gives you no way to set. The docs half of #4711 landed; the setter didn't.

### Why not incubate it first?

I seriously considered putting this in `api:incubator` behind `ExtendedTracerBuilder` etc., since that's the repo's standard de-risking move now and it didn't exist in 2022. I landed against it for one reason: the audience for scope attributes is instrumentation-library authors, and they're precisely the people who can't take an `-alpha` dependency without pushing it onto their users. The incubator earns its cost when the API *shape* is uncertain. This shape is spec-pinned and shipped in seven languages. Incubating would add six interfaces to defer a decision that's already been made elsewhere.

Happy to be argued out of this one — it's the main fork in the road.

---

## What the change does

### 1. Three `default` methods on the stable interfaces

`default` matters here. #4711 declared `LoggerBuilder.setAttributes` abstract, which was fine when the logs API was alpha — it's stable as of `@since 1.27.0` now, so that same line would be a breaking change japicmp rejects. All three need a default body.

The default returning `this` also means every existing implementor keeps working untouched, including the noop builders, the incubator's `ExtendedDefault*` classes, and anything out in the wild (the Java agent, Micrometer bridges, test doubles).

I checked every in-repo implementor, specifically looking for delegating wrappers that would silently swallow the new setter. There are none — everything is either one of the three SDK builders below or a noop where inheriting the default is exactly right.

### 2. Three SDK builders pass the attributes through

`ComponentRegistry.get()` already accepted an `Attributes` parameter and already threaded it into `InstrumentationScopeInfo`; `InstrumentationScopeInfoBuilder.setAttributes` already existed; OTLP already serialized them. The SDK builders were just passing `Attributes.empty()`. So this part is genuinely a three-line-per-file change.

### 3. `ComponentRegistry` stops silently swallowing conflicting attributes

This is the only piece that's more than plumbing, and the only piece you might reasonably call scope creep. My argument for bundling it:

The collision branch is **unreachable today**. Every SDK builder hardcodes `Attributes.empty()`, so two callers can never disagree. Adding the setter is what makes it live. Shipping the setter without this would mean one release with a reachable, completely silent footgun, fixed in the next.

Behaviour stays first-wins — that's spec-correct, since [spec#2789 "Remove scope attributes from the Scope identity"](https://github.com/open-telemetry/opentelemetry-specification/pull/2789) settled that attributes aren't part of identity, and this repo already implements that at the registry-keying layer. What changes is that you now find out:

```
Instrumentation scope [name=test_name, version=version, schemaUrl=http://schema.com] already
exists with attributes {k1="v1"}. The attributes {k1="v2"} provided by this call are ignored,
because instrumentation scope attributes are not part of scope identity. Obtain this scope with
consistent attributes to avoid this.
```

The warning is deliberately narrow — it only fires when *the caller* supplied non-empty attributes that differ. A caller who set nothing has had nothing taken away, so the very common mixed pattern (`getTracer("x")` somewhere, `tracerBuilder("x").setAttributes(...)` somewhere else) stays quiet. It goes through `ThrottlingLogger`, so a pathological loop can't flood anyone's logs.

Implementation-wise the four nested maps now hold a small `ScopeComponent` pairing the component with the `InstrumentationScopeInfo` it was built from — that's what makes the comparison possible on a cache hit. One extra object per distinct scope, not per call.

---

## Two calls you may want to overrule

I'd rather flag these than have them found in review.

**The null guard degrades instead of throwing.** `setAttributes(null)` logs via `ApiUsageLogger` and falls back to `Attributes.empty()`. This follows [api-design.md](docs/knowledge/api-design.md)'s instrumentation-boundary rule — throwing would crash an application because of a bug in an instrumentation library. The wrinkle is that the sibling methods (`setSchemaUrl`, `setInstrumentationVersion`) have no guard at all, so this is asymmetric. Without a guard, a null flows to `InstrumentationScopeInfo.create`'s `requireNonNull` and NPEs at `build()` time, far from the call site — which is why I added it. Easy to drop if you prefer consistency with the neighbours.

**No `@since` tags.** Deliberate — per [general-patterns.md](docs/knowledge/general-patterns.md) the release process adds those, they shouldn't be in the change.

---

## Verification

- `./gradlew build` — **BUILD SUCCESSFUL, 962 tasks, exit 0.** Every module: all test suites, japicmp, Checkstyle, Error Prone, NullAway, Spotless, AnimalSniffer, OSGi bundling.
- japicmp accepts all three new methods as non-breaking, and `opentelemetry-api.txt` is the only apidiff file that changes — which matters, because CI fails a PR when the build regenerates an apidiff that wasn't committed.
- Branch was at `origin/main` exactly (0 behind), so the japicmp baseline isn't polluted by upstream drift. Worth re-checking if you sit on this for a while — [build.md](docs/knowledge/build.md) calls out that specific trap.

One Error Prone catch worth knowing about if you tweak the comparison: my first version used `existing == requested` as a fast path and `ReferenceEquality` rejected it under `-Werror`. `requested.isEmpty()` short-circuits the common case just as well.

---

## The diff

You can apply the whole thing with `git apply`, or just paste file by file.

### API — the three interfaces

```diff
--- a/api/all/src/main/java/io/opentelemetry/api/trace/TracerBuilder.java
+++ b/api/all/src/main/java/io/opentelemetry/api/trace/TracerBuilder.java
@@ -5,6 +5,8 @@
 
 package io.opentelemetry.api.trace;
 
+import io.opentelemetry.api.common.Attributes;
+
 /**
  * Builder class for creating {@link Tracer} instances.
  *
@@ -34,6 +36,24 @@ public interface TracerBuilder {
    */
   TracerBuilder setInstrumentationVersion(String instrumentationScopeVersion);
 
+  /**
+   * Sets the instrumentation scope attributes of the resulting {@link Tracer}. Attributes are
+   * <i>not</i> part of {@link Tracer} identity.
+   *
+   * <p>Because attributes are not part of identity, obtaining multiple {@link Tracer}s with the
+   * same name, version, and schema URL but different attributes is not advised: the attributes of
+   * the first {@link Tracer} obtained are retained, and those of subsequent calls are ignored.
+   *
+   * <p>The default implementation is a no-op, retained for backwards compatibility with
+   * implementations written before this method existed.
+   *
+   * @param attributes The instrumentation scope attributes.
+   * @return this
+   */
+  default TracerBuilder setAttributes(Attributes attributes) {
+    return this;
+  }
+
   /**
    * Gets or creates a {@link Tracer} instance.
    *
```

`MeterBuilder` and `LoggerBuilder` are the same shape — same import, same method in the same position, with `Tracer`/`TracerBuilder` swapped for `Meter`/`MeterBuilder` and `Logger`/`LoggerBuilder`:

```diff
--- a/api/all/src/main/java/io/opentelemetry/api/metrics/MeterBuilder.java
+++ b/api/all/src/main/java/io/opentelemetry/api/metrics/MeterBuilder.java
@@ -5,6 +5,8 @@
 
 package io.opentelemetry.api.metrics;
 
+import io.opentelemetry.api.common.Attributes;
+
 /**
  * Builder class for creating {@link Meter} instances.
  *
@@ -34,6 +36,24 @@ public interface MeterBuilder {
    */
   MeterBuilder setInstrumentationVersion(String instrumentationScopeVersion);
 
+  /**
+   * Sets the instrumentation scope attributes of the resulting {@link Meter}. Attributes are
+   * <i>not</i> part of {@link Meter} identity.
+   *
+   * <p>Because attributes are not part of identity, obtaining multiple {@link Meter}s with the same
+   * name, version, and schema URL but different attributes is not advised: the attributes of the
+   * first {@link Meter} obtained are retained, and those of subsequent calls are ignored.
+   *
+   * <p>The default implementation is a no-op, retained for backwards compatibility with
+   * implementations written before this method existed.
+   *
+   * @param attributes The instrumentation scope attributes.
+   * @return this
+   */
+  default MeterBuilder setAttributes(Attributes attributes) {
+    return this;
+  }
+
   /**
    * Gets or creates a {@link Meter} instance.
    *
```

```diff
--- a/api/all/src/main/java/io/opentelemetry/api/logs/LoggerBuilder.java
+++ b/api/all/src/main/java/io/opentelemetry/api/logs/LoggerBuilder.java
@@ -5,6 +5,8 @@
 
 package io.opentelemetry.api.logs;
 
+import io.opentelemetry.api.common.Attributes;
+
 /**
  * Builder class for creating {@link Logger} instances.
  *
@@ -34,6 +36,24 @@ public interface LoggerBuilder {
    */
   LoggerBuilder setInstrumentationVersion(String instrumentationScopeVersion);
 
+  /**
+   * Sets the instrumentation scope attributes of the resulting {@link Logger}. Attributes are
+   * <i>not</i> part of {@link Logger} identity.
+   *
+   * <p>Because attributes are not part of identity, obtaining multiple {@link Logger}s with the
+   * same name, version, and schema URL but different attributes is not advised: the attributes of
+   * the first {@link Logger} obtained are retained, and those of subsequent calls are ignored.
+   *
+   * <p>The default implementation is a no-op, retained for backwards compatibility with
+   * implementations written before this method existed.
+   *
+   * @param attributes The instrumentation scope attributes.
+   * @return this
+   */
+  default LoggerBuilder setAttributes(Attributes attributes) {
+    return this;
+  }
+
   /**
    * Gets or creates a {@link Logger} instance.
    *
```

### SDK — the three builders

```diff
--- a/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkTracerBuilder.java
+++ b/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkTracerBuilder.java
@@ -8,6 +8,7 @@ package io.opentelemetry.sdk.trace;
 import io.opentelemetry.api.common.Attributes;
 import io.opentelemetry.api.trace.Tracer;
 import io.opentelemetry.api.trace.TracerBuilder;
+import io.opentelemetry.common.impl.ApiUsageLogger;
 import io.opentelemetry.sdk.common.internal.ComponentRegistry;
 import javax.annotation.Nullable;
 
@@ -17,6 +18,7 @@ class SdkTracerBuilder implements TracerBuilder {
   private final String instrumentationScopeName;
   @Nullable private String instrumentationScopeVersion;
   @Nullable private String schemaUrl;
+  private Attributes attributes = Attributes.empty();
 
   SdkTracerBuilder(ComponentRegistry<SdkTracer> registry, String instrumentationScopeName) {
     this.registry = registry;
@@ -35,9 +37,19 @@ class SdkTracerBuilder implements TracerBuilder {
     return this;
   }
 
+  @Override
+  public TracerBuilder setAttributes(Attributes attributes) {
+    if (attributes == null) {
+      ApiUsageLogger.logNullParam(TracerBuilder.class, "setAttributes", "attributes");
+      return this;
+    }
+    this.attributes = attributes;
+    return this;
+  }
+
   @Override
   public Tracer build() {
     return registry.get(
-        instrumentationScopeName, instrumentationScopeVersion, schemaUrl, Attributes.empty());
+        instrumentationScopeName, instrumentationScopeVersion, schemaUrl, attributes);
   }
 }
```

```diff
--- a/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterBuilder.java
+++ b/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterBuilder.java
@@ -8,6 +8,7 @@ package io.opentelemetry.sdk.metrics;
 import io.opentelemetry.api.common.Attributes;
 import io.opentelemetry.api.metrics.Meter;
 import io.opentelemetry.api.metrics.MeterBuilder;
+import io.opentelemetry.common.impl.ApiUsageLogger;
 import io.opentelemetry.sdk.common.internal.ComponentRegistry;
 import javax.annotation.Nullable;
 
@@ -17,6 +18,7 @@ class SdkMeterBuilder implements MeterBuilder {
   private final String instrumentationScopeName;
   @Nullable private String instrumentationScopeVersion;
   @Nullable private String schemaUrl;
+  private Attributes attributes = Attributes.empty();
 
   SdkMeterBuilder(ComponentRegistry<SdkMeter> registry, String instrumentationScopeName) {
     this.registry = registry;
@@ -35,9 +37,19 @@ class SdkMeterBuilder implements MeterBuilder {
     return this;
   }
 
+  @Override
+  public MeterBuilder setAttributes(Attributes attributes) {
+    if (attributes == null) {
+      ApiUsageLogger.logNullParam(MeterBuilder.class, "setAttributes", "attributes");
+      return this;
+    }
+    this.attributes = attributes;
+    return this;
+  }
+
   @Override
   public Meter build() {
     return registry.get(
-        instrumentationScopeName, instrumentationScopeVersion, schemaUrl, Attributes.empty());
+        instrumentationScopeName, instrumentationScopeVersion, schemaUrl, attributes);
   }
 }
```

```diff
--- a/sdk/logs/src/main/java/io/opentelemetry/sdk/logs/SdkLoggerBuilder.java
+++ b/sdk/logs/src/main/java/io/opentelemetry/sdk/logs/SdkLoggerBuilder.java
@@ -7,6 +7,7 @@ package io.opentelemetry.sdk.logs;
 
 import io.opentelemetry.api.common.Attributes;
 import io.opentelemetry.api.logs.LoggerBuilder;
+import io.opentelemetry.common.impl.ApiUsageLogger;
 import io.opentelemetry.sdk.common.internal.ComponentRegistry;
 import javax.annotation.Nullable;
 
@@ -16,6 +17,7 @@ final class SdkLoggerBuilder implements LoggerBuilder {
   private final String instrumentationScopeName;
   @Nullable private String instrumentationScopeVersion;
   @Nullable private String schemaUrl;
+  private Attributes attributes = Attributes.empty();
 
   SdkLoggerBuilder(ComponentRegistry<SdkLogger> registry, String instrumentationScopeName) {
     this.registry = registry;
@@ -34,9 +36,19 @@ final class SdkLoggerBuilder implements LoggerBuilder {
     return this;
   }
 
+  @Override
+  public SdkLoggerBuilder setAttributes(Attributes attributes) {
+    if (attributes == null) {
+      ApiUsageLogger.logNullParam(LoggerBuilder.class, "setAttributes", "attributes");
+      return this;
+    }
+    this.attributes = attributes;
+    return this;
+  }
+
   @Override
   public SdkLogger build() {
     return registry.get(
-        instrumentationScopeName, instrumentationScopeVersion, schemaUrl, Attributes.empty());
+        instrumentationScopeName, instrumentationScopeVersion, schemaUrl, attributes);
   }
 }
```

### SDK — `ComponentRegistry`

The unified diff for this one is noisy, because changing the map value type from `V` to `ScopeComponent<V>` touches every declaration and every `computeIfAbsent`. The whole file is easier to read and to paste.

<details>
<summary><b>Full file:</b> <code>sdk/common/src/main/java/io/opentelemetry/sdk/common/internal/ComponentRegistry.java</code></summary>

```java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.sdk.common.internal;

import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.internal.GuardedBy;
import io.opentelemetry.sdk.common.InstrumentationScopeInfo;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.annotation.Nullable;

/**
 * Component (tracer, meter, etc) registry class for all the provider classes (TracerProvider,
 * MeterProvider, etc.).
 *
 * <p>Components are identified by name, version, and schema. Name is required, but version and
 * schema are optional. Therefore, we have 4 possible scenarios for component keys:
 *
 * <ol>
 *   <li>Only name is provided, represented by {@link #componentByName}
 *   <li>Name and version are provided, represented by {@link #componentByNameAndVersion}
 *   <li>Name and schema are provided, represented by {@link #componentByNameAndSchema}
 *   <li>Name, version and schema are provided, represented by {@link
 *       #componentByNameVersionAndSchema}
 * </ol>
 *
 * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
 * at any time.
 *
 * @param <V> the type of the registered value.
 */
public final class ComponentRegistry<V> {

  private static final Logger internalLogger = Logger.getLogger(ComponentRegistry.class.getName());

  private final ThrottlingLogger logger = new ThrottlingLogger(internalLogger);

  private final Map<String, ScopeComponent<V>> componentByName = new ConcurrentHashMap<>();
  private final Map<String, Map<String, ScopeComponent<V>>> componentByNameAndVersion =
      new ConcurrentHashMap<>();
  private final Map<String, Map<String, ScopeComponent<V>>> componentByNameAndSchema =
      new ConcurrentHashMap<>();
  private final Map<String, Map<String, Map<String, ScopeComponent<V>>>>
      componentByNameVersionAndSchema = new ConcurrentHashMap<>();

  private final Object lock = new Object();

  @GuardedBy("lock")
  private final Set<V> allComponents = Collections.newSetFromMap(new IdentityHashMap<>());

  private final Function<InstrumentationScopeInfo, V> factory;

  public ComponentRegistry(Function<InstrumentationScopeInfo, V> factory) {
    this.factory = factory;
  }

  /**
   * Returns the component associated with the {@code name}, {@code version}, and {@code schemaUrl}.
   * {@link Attributes} are not part of component identity. When a component with the same {@code
   * name}, {@code version}, and {@code schemaUrl} already exists, the {@link Attributes} it was
   * first created with are retained and the {@code attributes} of this call are ignored. A warning
   * is logged when this discards a non-empty set of {@code attributes}.
   */
  public V get(
      String name, @Nullable String version, @Nullable String schemaUrl, Attributes attributes) {
    ScopeComponent<V> scopeComponent;
    if (version != null && schemaUrl != null) {
      Map<String, Map<String, ScopeComponent<V>>> componentByVersionAndSchema =
          componentByNameVersionAndSchema.computeIfAbsent(
              name, unused -> new ConcurrentHashMap<>());
      Map<String, ScopeComponent<V>> componentBySchema =
          componentByVersionAndSchema.computeIfAbsent(version, unused -> new ConcurrentHashMap<>());
      scopeComponent =
          componentBySchema.computeIfAbsent(
              schemaUrl,
              schemaUrl1 ->
                  buildComponent(
                      InstrumentationScopeInfo.builder(name)
                          .setVersion(version)
                          .setSchemaUrl(schemaUrl1)
                          .setAttributes(attributes)
                          .build()));
    } else if (version != null) { // schemaUrl == null
      Map<String, ScopeComponent<V>> componentByVersion =
          componentByNameAndVersion.computeIfAbsent(name, unused -> new ConcurrentHashMap<>());
      scopeComponent =
          componentByVersion.computeIfAbsent(
              version,
              version1 ->
                  buildComponent(
                      InstrumentationScopeInfo.builder(name)
                          .setVersion(version1)
                          .setAttributes(attributes)
                          .build()));
    } else if (schemaUrl != null) { // version == null
      Map<String, ScopeComponent<V>> componentBySchema =
          componentByNameAndSchema.computeIfAbsent(name, unused -> new ConcurrentHashMap<>());
      scopeComponent =
          componentBySchema.computeIfAbsent(
              schemaUrl,
              schemaUrl1 ->
                  buildComponent(
                      InstrumentationScopeInfo.builder(name)
                          .setSchemaUrl(schemaUrl1)
                          .setAttributes(attributes)
                          .build()));
    } else { // schemaUrl == null && version == null
      scopeComponent =
          componentByName.computeIfAbsent(
              name,
              name1 ->
                  buildComponent(
                      InstrumentationScopeInfo.builder(name1).setAttributes(attributes).build()));
    }

    logIfAttributesIgnored(scopeComponent.scopeInfo, attributes);
    return scopeComponent.component;
  }

  /**
   * Logs a warning when {@code requested} attributes are non-empty and differ from the attributes
   * the scope was first created with, since the requested attributes are silently discarded.
   */
  private void logIfAttributesIgnored(
      InstrumentationScopeInfo existingScopeInfo, Attributes requested) {
    Attributes existing = existingScopeInfo.getAttributes();
    // Fast path: callers which requested no attributes have nothing discarded, and this covers the
    // overwhelmingly common case of a caller which never set attributes at all.
    if (requested.isEmpty() || existing.equals(requested)) {
      return;
    }
    logger.log(
        Level.WARNING,
        "Instrumentation scope "
            + scopeIdentity(existingScopeInfo)
            + " already exists with attributes "
            + existing
            + ". The attributes "
            + requested
            + " provided by this call are ignored, because instrumentation scope attributes are not"
            + " part of scope identity. Obtain this scope with consistent attributes to avoid"
            + " this.");
  }

  private static String scopeIdentity(InstrumentationScopeInfo scopeInfo) {
    return "[name="
        + scopeInfo.getName()
        + ", version="
        + scopeInfo.getVersion()
        + ", schemaUrl="
        + scopeInfo.getSchemaUrl()
        + "]";
  }

  private ScopeComponent<V> buildComponent(InstrumentationScopeInfo instrumentationScopeInfo) {
    V component = factory.apply(instrumentationScopeInfo);
    synchronized (lock) {
      allComponents.add(component);
    }
    return new ScopeComponent<>(instrumentationScopeInfo, component);
  }

  /**
   * Returns a {@code Collection} view of the registered components.
   *
   * @return a {@code Collection} view of the registered components.
   */
  public Collection<V> getComponents() {
    synchronized (lock) {
      return Collections.unmodifiableCollection(new ArrayList<>(allComponents));
    }
  }

  /**
   * A registered component, paired with the {@link InstrumentationScopeInfo} it was created with so
   * that later calls can detect when they supply conflicting scope attributes.
   */
  private static final class ScopeComponent<V> {
    private final InstrumentationScopeInfo scopeInfo;
    private final V component;

    private ScopeComponent(InstrumentationScopeInfo scopeInfo, V component) {
      this.scopeInfo = scopeInfo;
      this.component = component;
    }
  }
}
```

</details>

One small thing I tidied while in there: the original had `} else if (version != null) { ... }` followed by a fresh `if (schemaUrl != null)`. That second `if` was only reachable when `version == null`, so it's now an `else if` in the same chain. Same behaviour, one less thing to reason about — say the word if you'd rather I left it alone.

### New test file

<details>
<summary><b>New file:</b> <code>api/all/src/test/java/io/opentelemetry/api/ScopeAttributesTest.java</code></summary>

```java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.api;

import static org.assertj.core.api.Assertions.assertThat;

import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.logs.LoggerBuilder;
import io.opentelemetry.api.logs.LoggerProvider;
import io.opentelemetry.api.metrics.MeterBuilder;
import io.opentelemetry.api.metrics.MeterProvider;
import io.opentelemetry.api.trace.TracerBuilder;
import io.opentelemetry.api.trace.TracerProvider;
import org.junit.jupiter.api.Test;

/**
 * Verifies that the default {@code setAttributes} implementations are no-ops which preserve builder
 * chaining, and that no-op providers accept scope attributes without failing.
 */
class ScopeAttributesTest {

  private static final Attributes ATTRIBUTES = Attributes.builder().put("key", "value").build();

  @Test
  void noopTracerBuilder_AcceptsAttributes() {
    TracerBuilder builder = TracerProvider.noop().tracerBuilder("test");

    assertThat(builder.setAttributes(ATTRIBUTES)).isSameAs(builder);
    assertThat(
            builder
                .setInstrumentationVersion("1.0")
                .setSchemaUrl("http://url")
                .setAttributes(ATTRIBUTES)
                .build())
        .isSameAs(TracerProvider.noop().get("test"));
  }

  @Test
  void noopMeterBuilder_AcceptsAttributes() {
    MeterBuilder builder = MeterProvider.noop().meterBuilder("test");

    assertThat(builder.setAttributes(ATTRIBUTES)).isSameAs(builder);
    assertThat(
            builder
                .setInstrumentationVersion("1.0")
                .setSchemaUrl("http://url")
                .setAttributes(ATTRIBUTES)
                .build())
        .isSameAs(MeterProvider.noop().get("test"));
  }

  @Test
  void noopLoggerBuilder_AcceptsAttributes() {
    LoggerBuilder builder = LoggerProvider.noop().loggerBuilder("test");

    assertThat(builder.setAttributes(ATTRIBUTES)).isSameAs(builder);
    assertThat(
            builder
                .setInstrumentationVersion("1.0")
                .setSchemaUrl("http://url")
                .setAttributes(ATTRIBUTES)
                .build())
        .isSameAs(LoggerProvider.noop().get("test"));
  }
}
```

</details>

### Existing tests

<details>
<summary><code>sdk/common/.../ComponentRegistryTest.java</code> — first-wins, warning fires, warning stays quiet</summary>

```diff
@@ -8,7 +8,10 @@ package io.opentelemetry.sdk.common.internal;
 import static java.util.stream.Collectors.joining;
 import static org.assertj.core.api.Assertions.assertThat;
 
+import io.github.netmikey.logunit.api.LogCapturer;
 import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.internal.testing.slf4j.SuppressLogger;
+import io.opentelemetry.sdk.common.InstrumentationScopeInfo;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Random;
@@ -16,8 +19,10 @@ import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.IntStream;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
 
 class ComponentRegistryTest {
 
@@ -28,7 +33,11 @@ class ComponentRegistryTest {
   private final ComponentRegistry<TestComponent> registry =
       new ComponentRegistry<>(unused -> new TestComponent());
 
+  @RegisterExtension
+  LogCapturer logs = LogCapturer.create().captureForType(ComponentRegistry.class);
+
   @Test
+  @SuppressLogger(ComponentRegistry.class)
   void get_SameInstance() {
     assertThat(registry.get(NAME, null, null, Attributes.empty()))
         .isSameAs(registry.get(NAME, null, null, Attributes.empty()))
@@ -61,6 +70,51 @@ class ComponentRegistryTest {
         .isNotSameAs(registry.get(NAME, null, null, Attributes.empty()));
   }
 
+  @Test
+  @SuppressLogger(ComponentRegistry.class)
+  void get_FirstAttributesWin() {
+    AtomicReference<InstrumentationScopeInfo> scopeInfo = new AtomicReference<>();
+    ComponentRegistry<TestComponent> registry =
+        new ComponentRegistry<>(
+            info -> {
+              scopeInfo.set(info);
+              return new TestComponent();
+            });
+
+    registry.get(NAME, VERSION, SCHEMA_URL, ATTRIBUTES);
+    registry.get(NAME, VERSION, SCHEMA_URL, Attributes.builder().put("k1", "v2").build());
+
+    assertThat(scopeInfo.get().getAttributes()).isEqualTo(ATTRIBUTES);
+  }
+
+  @Test
+  void get_DifferentAttributesLogged() {
+    registry.get(NAME, VERSION, SCHEMA_URL, ATTRIBUTES);
+    registry.get(NAME, VERSION, SCHEMA_URL, Attributes.builder().put("k1", "v2").build());
+
+    assertThat(logs.getEvents()).hasSize(1);
+    assertThat(logs.getEvents().get(0).getMessage())
+        .contains("already exists with attributes")
+        .contains(NAME)
+        .contains(VERSION)
+        .contains(SCHEMA_URL);
+  }
+
+  @Test
+  void get_SameOrEmptyAttributesNotLogged() {
+    registry.get(NAME, VERSION, SCHEMA_URL, ATTRIBUTES);
+    // Identical attributes discard nothing.
+    registry.get(NAME, VERSION, SCHEMA_URL, ATTRIBUTES);
+    registry.get(NAME, VERSION, SCHEMA_URL, Attributes.builder().put("k1", "v1").build());
+    // A caller which requested no attributes has nothing discarded.
+    registry.get(NAME, VERSION, SCHEMA_URL, Attributes.empty());
+    // Scopes which have no attributes at all.
+    registry.get(NAME + "_1", null, null, Attributes.empty());
+    registry.get(NAME + "_1", null, null, Attributes.empty());
+
+    assertThat(logs.getEvents()).isEmpty();
+  }
+
   @Test
   @SuppressWarnings("ReturnValueIgnored")
   void getComponents_HighConcurrency() throws ExecutionException, InterruptedException {
```

`get_SameInstance` deliberately calls with differing attributes, so it now needs `@SuppressLogger` to keep the test output clean.

</details>

<details>
<summary><code>sdk/trace/.../SdkTracerProviderTest.java</code></summary>

```diff
@@ -22,6 +22,7 @@ import io.opentelemetry.internal.testing.slf4j.SuppressLogger;
 import io.opentelemetry.sdk.common.Clock;
 import io.opentelemetry.sdk.common.CompletableResultCode;
 import io.opentelemetry.sdk.common.InstrumentationScopeInfo;
+import io.opentelemetry.sdk.common.internal.ComponentRegistry;
 import io.opentelemetry.sdk.common.internal.ExceptionAttributeResolver;
 import io.opentelemetry.sdk.common.internal.ScopeConfigurator;
 import io.opentelemetry.sdk.resources.Resource;
@@ -183,16 +184,45 @@ class SdkTracerProviderTest {
         InstrumentationScopeInfo.builder("theName")
             .setVersion("theVersion")
             .setSchemaUrl("http://url")
+            .setAttributes(Attributes.builder().put("key", "value").build())
             .build();
     Tracer tracer =
         tracerProvider
             .tracerBuilder(expected.getName())
             .setInstrumentationVersion(expected.getVersion())
             .setSchemaUrl(expected.getSchemaUrl())
+            .setAttributes(expected.getAttributes())
             .build();
     assertThat(((SdkTracer) tracer).getInstrumentationScopeInfo()).isEqualTo(expected);
   }
 
+  @Test
+  @SuppressLogger(ComponentRegistry.class)
+  void scopeAttributes_NotPartOfTracerIdentity() {
+    Attributes attributes = Attributes.builder().put("key", "value").build();
+    Tracer tracer = tracerProvider.tracerBuilder("test").setAttributes(attributes).build();
+
+    // Same name, version, and schema URL resolves to the same tracer regardless of attributes, and
+    // the attributes of the first tracer obtained are retained.
+    assertThat(tracerProvider.tracerBuilder("test").build()).isSameAs(tracer);
+    assertThat(
+            tracerProvider
+                .tracerBuilder("test")
+                .setAttributes(Attributes.builder().put("key", "other").build())
+                .build())
+        .isSameAs(tracer);
+    assertThat(((SdkTracer) tracer).getInstrumentationScopeInfo().getAttributes())
+        .isEqualTo(attributes);
+  }
+
+  @Test
+  void scopeAttributes_Null() {
+    Tracer tracer = tracerProvider.tracerBuilder("test").setAttributes(null).build();
+
+    assertThat(((SdkTracer) tracer).getInstrumentationScopeInfo().getAttributes())
+        .isEqualTo(Attributes.empty());
+  }
+
   @Test
   void propagatesEnablementToTracerDirectly() {
     propagatesEnablementToTracer(true);
```

</details>

<details>
<summary><code>sdk/metrics/.../SdkMeterProviderTest.java</code> — includes an end-to-end check that attributes reach collected <code>MetricData</code></summary>

```diff
@@ -26,8 +26,10 @@ import io.opentelemetry.api.metrics.ObservableLongCounter;
 import io.opentelemetry.api.metrics.ObservableLongMeasurement;
 import io.opentelemetry.context.Context;
 import io.opentelemetry.context.Scope;
+import io.opentelemetry.internal.testing.slf4j.SuppressLogger;
 import io.opentelemetry.sdk.common.CompletableResultCode;
 import io.opentelemetry.sdk.common.InstrumentationScopeInfo;
+import io.opentelemetry.sdk.common.internal.ComponentRegistry;
 import io.opentelemetry.sdk.common.internal.ScopeConfigurator;
 import io.opentelemetry.sdk.metrics.data.MetricData;
 import io.opentelemetry.sdk.metrics.export.MetricReader;
@@ -91,6 +93,73 @@ class SdkMeterProviderTest {
     assertThat(sdkMeterProvider.get(null)).isSameAs(sdkMeterProvider.get("unknown"));
   }
 
+  @Test
+  void meterBuilder_ScopeAttributesPropagateToMetrics() {
+    InMemoryMetricReader sdkMeterReader = InMemoryMetricReader.create();
+    SdkMeterProvider sdkMeterProvider =
+        sdkMeterProviderBuilder.registerMetricReader(sdkMeterReader).build();
+    InstrumentationScopeInfo expected =
+        InstrumentationScopeInfo.builder("test")
+            .setVersion("version")
+            .setSchemaUrl("http://url")
+            .setAttributes(Attributes.builder().put("key", "value").build())
+            .build();
+
+    sdkMeterProvider
+        .meterBuilder("test")
+        .setInstrumentationVersion("version")
+        .setSchemaUrl("http://url")
+        .setAttributes(Attributes.builder().put("key", "value").build())
+        .build()
+        .counterBuilder("counter")
+        .build()
+        .add(1);
+
+    assertThat(sdkMeterReader.collectAllMetrics())
+        .satisfiesExactly(metricData -> assertThat(metricData).hasInstrumentationScope(expected));
+  }
+
+  @Test
+  @SuppressLogger(ComponentRegistry.class)
+  void meterBuilder_ScopeAttributesNotPartOfIdentity() {
+    SdkMeterProvider sdkMeterProvider =
+        sdkMeterProviderBuilder.registerMetricReader(InMemoryMetricReader.create()).build();
+    Meter meter =
+        sdkMeterProvider
+            .meterBuilder("test")
+            .setAttributes(Attributes.builder().put("key", "value").build())
+            .build();
+
+    assertThat(sdkMeterProvider.meterBuilder("test").build()).isSameAs(meter);
+    assertThat(
+            sdkMeterProvider
+                .meterBuilder("test")
+                .setAttributes(Attributes.builder().put("key", "other").build())
+                .build())
+        .isSameAs(meter);
+  }
+
+  @Test
+  void meterBuilder_ScopeAttributesNull() {
+    InMemoryMetricReader sdkMeterReader = InMemoryMetricReader.create();
+    SdkMeterProvider sdkMeterProvider =
+        sdkMeterProviderBuilder.registerMetricReader(sdkMeterReader).build();
+
+    sdkMeterProvider
+        .meterBuilder("test")
+        .setAttributes(null)
+        .build()
+        .counterBuilder("counter")
+        .build()
+        .add(1);
+
+    assertThat(sdkMeterReader.collectAllMetrics())
+        .satisfiesExactly(
+            metricData ->
+                assertThat(metricData)
+                    .hasInstrumentationScope(InstrumentationScopeInfo.create("test")));
+  }
+
   @Test
   void collectAllSyncInstruments() {
     InMemoryMetricReader sdkMeterReader = InMemoryMetricReader.create();
```

</details>

<details>
<summary><code>sdk/logs/.../SdkLoggerProviderTest.java</code></summary>

```diff
@@ -28,6 +28,7 @@ import io.opentelemetry.internal.testing.slf4j.SuppressLogger;
 import io.opentelemetry.sdk.common.Clock;
 import io.opentelemetry.sdk.common.CompletableResultCode;
 import io.opentelemetry.sdk.common.InstrumentationScopeInfo;
+import io.opentelemetry.sdk.common.internal.ComponentRegistry;
 import io.opentelemetry.sdk.common.internal.ScopeConfigurator;
 import io.opentelemetry.sdk.logs.data.LogRecordData;
 import io.opentelemetry.sdk.logs.internal.LoggerConfig;
@@ -193,6 +194,7 @@ class SdkLoggerProviderTest {
         InstrumentationScopeInfo.builder("test")
             .setVersion("version")
             .setSchemaUrl("http://url")
+            .setAttributes(Attributes.builder().put("key", "value").build())
             .build();
     assertThat(
             ((SdkLogger)
@@ -200,11 +202,38 @@ class SdkLoggerProviderTest {
                         .loggerBuilder("test")
                         .setInstrumentationVersion("version")
                         .setSchemaUrl("http://url")
+                        .setAttributes(Attributes.builder().put("key", "value").build())
                         .build())
                 .getInstrumentationScopeInfo())
         .isEqualTo(expected);
   }
 
+  @Test
+  @SuppressLogger(ComponentRegistry.class)
+  void loggerBuilder_ScopeAttributesNotPartOfIdentity() {
+    Attributes attributes = Attributes.builder().put("key", "value").build();
+    SdkLogger logger =
+        (SdkLogger) sdkLoggerProvider.loggerBuilder("test").setAttributes(attributes).build();
+
+    assertThat(sdkLoggerProvider.loggerBuilder("test").build()).isSameAs(logger);
+    assertThat(
+            sdkLoggerProvider
+                .loggerBuilder("test")
+                .setAttributes(Attributes.builder().put("key", "other").build())
+                .build())
+        .isSameAs(logger);
+    assertThat(logger.getInstrumentationScopeInfo().getAttributes()).isEqualTo(attributes);
+  }
+
+  @Test
+  void loggerBuilder_ScopeAttributesNull() {
+    assertThat(
+            ((SdkLogger) sdkLoggerProvider.loggerBuilder("test").setAttributes(null).build())
+                .getInstrumentationScopeInfo()
+                .getAttributes())
+        .isEqualTo(Attributes.empty());
+  }
+
   @Test
   void loggerBuilder_DefaultLoggerName() {
     assertThat(
```

</details>

### CHANGELOG and apidiff

The apidiff is generated — run `./gradlew jApiCmp` rather than hand-editing it.

<details>
<summary>CHANGELOG.md + docs/apidiffs/current_vs_latest/opentelemetry-api.txt</summary>

```diff
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,20 @@
 
 ## Unreleased
 
+### API
+
+* Add `setAttributes(Attributes)` to `TracerBuilder`, `MeterBuilder`, and `LoggerBuilder`, allowing
+  instrumentation to set instrumentation scope attributes. Scope attributes are not part of scope
+  identity: when a scope with the same name, version, and schema URL already exists, the attributes
+  it was first obtained with are retained
+  ([#4695](https://github.com/open-telemetry/opentelemetry-java/issues/4695))
+
+### SDK
+
+* Log a warning when a tracer, meter, or logger is obtained with instrumentation scope attributes
+  which differ from those the scope was first created with, since the requested attributes are
+  discarded ([#4695](https://github.com/open-telemetry/opentelemetry-java/issues/4695))
+
 ## Version 1.65.0 (2026-08-07)
```

```diff
--- a/docs/apidiffs/current_vs_latest/opentelemetry-api.txt
+++ b/docs/apidiffs/current_vs_latest/opentelemetry-api.txt
@@ -1,2 +1,10 @@
 Comparing source compatibility of opentelemetry-api-1.66.0-SNAPSHOT.jar against opentelemetry-api-1.64.0.jar
-No changes.
\ No newline at end of file
+***  MODIFIED INTERFACE: PUBLIC ABSTRACT io.opentelemetry.api.logs.LoggerBuilder  (not serializable)
+	===  CLASS FILE FORMAT VERSION: 52.0 <- 52.0
+	+++  NEW METHOD: PUBLIC(+) io.opentelemetry.api.logs.LoggerBuilder setAttributes(io.opentelemetry.api.common.Attributes)
+***  MODIFIED INTERFACE: PUBLIC ABSTRACT io.opentelemetry.api.metrics.MeterBuilder  (not serializable)
+	===  CLASS FILE FORMAT VERSION: 52.0 <- 52.0
+	+++  NEW METHOD: PUBLIC(+) io.opentelemetry.api.metrics.MeterBuilder setAttributes(io.opentelemetry.api.common.Attributes)
+***  MODIFIED INTERFACE: PUBLIC ABSTRACT io.opentelemetry.api.trace.TracerBuilder  (not serializable)
+	===  CLASS FILE FORMAT VERSION: 52.0 <- 52.0
+	+++  NEW METHOD: PUBLIC(+) io.opentelemetry.api.trace.TracerBuilder setAttributes(io.opentelemetry.api.common.Attributes)
```

</details>

---

## What's left

Housekeeping, none of it code:

- **CHANGELOG links the issue, not a PR.** Chicken-and-egg — swap in the PR number once one exists.
- **Compliance matrix.** Flipping Java's `-` to `+` is a separate change in `opentelemetry-specification`, after this merges.
- **Issue tidy-up.** The Zipkin and Jaeger checkboxes on #4695 can't be completed and should be struck, otherwise the issue can never close.
- **CLA.** The CNCF CLA bot gates submission; worth checking whether a corporate CLA already covers you.

## One suggestion, take it or leave it

Before opening anything, it may be worth a comment on #4695 laying out the compliance-matrix evidence and the proposed `default setAttributes(Attributes)` shape, and waiting for a maintainer to nod.

#4711 was working, tested code that sat for four months collecting stale-bot comments and then got closed over a design question raised after submission — piecemeal setters versus a `Scope` object. This proposal is the same shape as #4711, so exactly the same objection is available. A day's wait for an ack is cheap compared to another four-month round trip, and it has the side benefit of finally answering chukunx's question.

## Things I'd genuinely like a second opinion on

1. **`default` on the stable interfaces vs. incubating first.** My reasoning is above, but you know the maintainers' current appetite better than I do.
2. **The `ComponentRegistry` warning** — right call to bundle, or split it out?
3. **`WARNING` the right level** for that message, or too shouty for something the spec merely calls "not advised"?
4. **The null guard** — keep it, or match the unguarded neighbours?
